docs/core/deploying/trimming/trimming-concepts.md

```diff
@@ -0,0 +1,402 @@
+ms.author: angocke
+ms.date: 11/04/2025
+ai-usage: ai-assisted
+---
+# Understanding trim analysis
+
+This article explains the fundamental concepts behind trim analysis to help you understand why certain code patterns produce warnings and how to make your code trim-compatible. Understanding these concepts will help you make informed decisions when addressing trim warnings rather than simply "spreading attributes around to silence the tooling."
+
+## How the trimmer analyzes code
+
+The trimmer performs **static analysis** at publish time to determine which code is used by your application. It starts from known entry points (like your `Main` method) and follows the code paths through your application.
+
+### What the trimmer can understand
+
+The trimmer excels at analyzing direct, compile-time-visible code patterns:
+
+```csharp
+// The trimmer CAN understand these patterns:
+var date = new DateTime();
+date.AddDays(1);  // Direct method call - trimmer knows AddDays is used
+
+var list = new List<string>();
+list.Add("hello");  // Generic method call - trimmer knows List<string>.Add is used
+
+string result = MyUtility.Process("data");  // Direct static method call
+```
+
+In these examples, the trimmer can follow the code path and mark `DateTime.AddDays`, `List<string>.Add`, and `MyUtility.Process` as used code that should be kept in the final application.
+
+### What the trimmer cannot understand
+
+The trimmer struggles with dynamic operations where the target of an operation isn't known until runtime:
+
+```csharp
+// The trimmer CANNOT fully understand these patterns:
+Type type = Type.GetType(Console.ReadLine());  // Type name from user input
+type.GetMethod("SomeMethod");  // Which method? On which type?
+
+object obj = GetSomeObject();
+obj.GetType().GetProperties();  // What type will obj be at runtime?
+
+Assembly asm = Assembly.LoadFrom(pluginPath);  // What's in this assembly?
+```
+
+In these examples, the trimmer has no way to know:
+
+- Which type the user will enter
+- What type `GetSomeObject()` returns
+- What code exists in the dynamically loaded assembly
+
+This is the fundamental problem that trim warnings address.
+
+## The reflection problem
+
+Reflection allows code to inspect and invoke types and members dynamically at runtime. This is powerful but creates a challenge for static analysis.
+
+### Why reflection breaks trimming
+
+Consider this example:
+
+```csharp
+void PrintMethodNames(Type type)
+{
+    foreach (var method in type.GetMethods())
+    {
+        Console.WriteLine(method.Name);
+    }
+}
+
+// Called somewhere in the app
+PrintMethodNames(typeof(DateTime));
+```
+
+From the trimmer's perspective:
+
+1. It sees `type.GetMethods()` is called
+2. It doesn't know what `type` will be (it's a parameter)
+3. It can't determine which types' methods need to be preserved
+4. Without guidance, it might remove methods from `DateTime`, breaking the code
+
+This is why the trimmer produces a warning on this code.
```

Comment:
```suggestion
- It sees `type.GetMethods()` is called.
- It doesn't know what `type` will be (it's a parameter).
- It can't determine which types' methods need to be preserved.
- Without guidance, it might remove methods from `DateTime`, breaking the code.

Consequently, the trimmer produces a warning on this code.
```